syntax = "proto3";

// Define o pacote para o Java (para organizar as classes geradas)
option java_package = "org.fauxgartic.grpc";
// Define que queremos várias classes (uma por mensagem) ou uma classe "outer" gigante
option java_multiple_files = true;

// Serviço principal do Jogo
service GarticService {
  // O cliente se conecta e recebe um ID único e o estado atual
  rpc EntrarNoJogo (Jogador) returns (EstadoDoJogo);

  // O cliente abre um fluxo (stream) para RECEBER eventos do servidor continuamente
  // (Ex: outros desenhando, chat, troca de turno)
  rpc ReceberEventos (Jogador) returns (stream EventoDeJogo);

  // O cliente envia uma ação (desenho ou palpite) para o servidor
  rpc EnviarAcao (AcaoJogador) returns (Vazio);
}

// Mensagens de Dados

message Jogador {
  string nome = 1;
  string id = 2; // UUID gerado pelo cliente ou servidor
}

message EstadoDoJogo {
  bool sou_o_desenhista = 1; // Diz se eu começo desenhando
  string desenhista_atual = 2; // Nome de quem está desenhando
  string palavra_atual = 3;    // Se eu for desenhista, recebo a palavra. Se não, vazio.
}

message AcaoJogador {
  Jogador jogador = 1;
  oneof acao {
    Traco traco = 2;       // Se for desenhista enviando traço
    string palpite = 3;    // Se for adivinhador enviando texto
    bool limpar_tela = 4;  // Se o desenhista limpou a tela
  }
}

message Traco {
  double x = 1;
  double y = 2;
  string cor = 3;
  bool novo_traco = 4; // True se for um clique novo (START), False se for arrastar (DRAW)
}

// O que o servidor manda de volta para os clientes no stream
message EventoDeJogo {
  oneof evento {
    Traco desenho = 1;          // Servidor repassando um desenho de alguém
    string mensagem_chat = 2;   // Servidor mandando msg de chat/acerto
    Rodada mudanca_rodada = 3;  // Servidor avisando: "Agora é a vez de X"
    bool limpar_tela = 4;       // Servidor mandando limpar
  }
}

message Rodada {
  string nome_desenhista = 1;
  // Se quem recebe for o novo desenhista, isso terá a palavra. Senão, vazio.
  string palavra_secreta = 2; 
}

message Vazio {}